src/mlpack/bindings/go/print_doc_functions_impl.hpp

```diff
@@ -0,0 +1,541 @@
+  else
+    return "false";
+}
+
+/**
+ * Given a parameter name, print its corresponding default value.
+ */
+inline std::string PrintDefault(const std::string& paramName)
+{
+  if (CLI::Parameters().count(paramName) == 0)
+    throw std::invalid_argument("unknown parameter " + paramName + "!");
+
+  const util::ParamData& d = CLI::Parameters()[paramName];
+
+  std::string defaultValue;
+  CLI::GetSingleton().functionMap[d.tname]["DefaultParam"](d, NULL,
+      (void*) &defaultValue);
+
+  return defaultValue;
+}
+
+// Recursion base case.
+std::string PrintOptionalInputs() { return ""; }
+
+/**
+ * Print an input option.  This will throw an exception if the parameter does
+ * not exist in CLI.
+ */
+template<typename T, typename... Args>
+std::string PrintOptionalInputs(const std::string& paramName,
+                              const T& value,
+                              Args... args)
+{
+  // See if this is part of the program.
+  std::string result = "";
+  if (CLI::Parameters().count(paramName) > 0)
+  {
+    const util::ParamData& d = CLI::Parameters()[paramName];
+    if (d.input && !d.required)
+    {
+      std::string goParamName = CamelCase(paramName);
+
+      // Print the input option.
+      std::ostringstream oss;
+      oss << "  param." << goParamName << " = ";
+
+      // Special handling is needed for model types.
+      if (PrintDefault(paramName) == "nil")
+      {
+        oss << "&";
+        oss << CamelCase(PrintValue(value, d.tname == TYPENAME(std::string)));
+      }
+      else
+      {
+        oss << PrintValue(value, d.tname == TYPENAME(std::string));
+      }
+      oss << "\n";
+      result = oss.str();
+    }
+  }
+  else
+  {
+    // Unknown parameter!
+    throw std::runtime_error("Unknown parameter '" + paramName + "' " +
+        "encountered while assembling documentation!  Check PROGRAM_INFO() " +
+        "declaration.");
+  }
+
+  // Continue recursion.
+  std::string rest = PrintOptionalInputs(args...);
+  if (rest != "" && result != "")
+    result += rest;
+  else if (result == "")
+    result = rest;
+
+  return result;
+}
+
+// Recursion base case.
+std::string PrintInputOptions() { return ""; }
+
+/**
+ * Print an input option.  This will throw an exception if the parameter does
+ * not exist in CLI.
+ */
+template<typename T, typename... Args>
+std::string PrintInputOptions(const std::string& paramName,
+                              const T& value,
+                              Args... args)
+{
+  // See if this is part of the program.
+  std::string result = "";
+
+  if (CLI::Parameters().count(paramName) > 0)
+  {
+    const util::ParamData& d = CLI::Parameters()[paramName];
+    if (d.input && d.required)
+    {
+      // Print the input option.
+      std::ostringstream oss;
+
+      // Special handling is needed for model types.
+      if (PrintDefault(paramName) == "nil")
+      {
+        oss << "&";
+        oss << CamelCase(PrintValue(value, d.tname == TYPENAME(std::string)));
```

Comment:
Here too I'm not sure if we should camel-case it.  Here's an example where the source doesn't match the documentation:

http://mlpack.ratml.org/doc/mlpack-git/go_documentation.html#hmm_generate_detailed-documentation

(`hmm` vs. `Hmm`)

Reply:
Updated.